textpattern/vendors/Textpattern/Plugin/Plugin.php

```diff
@@ -145,6 +145,88 @@ public function install($plugin, $status = null)
         return $message;
     }
 
+    /**
+     * Verify plugin
+     *
+     * @param  string       $plugin   Plugin_base64
+     *
+     * @return string|array
+     */
+
+    public function verify($plugin)
+    {
+        // Check for pre-4.0 style plugin.
+        if (strpos($plugin, '$plugin=\'') !== false) {
+            // Try to increase PCRE's backtrack limit in PHP 5.2+ to accommodate to
+            // x-large plugins. See https://bugs.php.net/bug.php?id=40846.
+            @ini_set('pcre.backtrack_limit', '1000000');
+            $plugin = preg_replace('@.*\$plugin=\'([\w=+/]+)\'.*@s', '$1', $plugin);
+            // Have we hit yet another PCRE restriction?
+            if ($plugin === null) {
+
+                return array(gTxt('plugin_pcre_error', array('{errno}' => preg_last_error())), E_ERROR);
+            }
+        }
+
+        // Strip out #comment lines.
+        $plugin = preg_replace('/^#.*$/m', '', $plugin);
+
+        if ($plugin === null) {
+
+            return array(gTxt('plugin_pcre_error', array('{errno}' => preg_last_error())), E_ERROR);
+        }
+
+        if (isset($plugin)) {
+            $plugin_encoded = $plugin;
+            $plugin = base64_decode($plugin);
+
+            if (strncmp($plugin, "\x1F\x8B", 2) === 0) {
+                if (function_exists('gzinflate')) {
+                    $plugin = gzinflate(substr($plugin, 10));
+                } else {
+
+                    return array(gTxt('plugin_compression_unsupported'), E_ERROR);
+                }
+            }
+
+            if ($plugin = @unserialize($plugin)) {
```

Comment:
This isn't really an "API" call if it renders stuff to the browser. Could we split this function here and call it, I dunno, `Plugin->unpack` or something? Or even keep it as `verify` I don't mind. Maybe make it `throw` error messages properly, that we can `catch` externally, or if everything succeeds, it should return the array of unserialized content.

Then we can defer the rendering portion back to `txp_plugin.php` for the time being. Bonuses:

1. Separation of data processing and display.
2. Plugin authors get to use the API to check / unpack plugins without UI baggage.

Longer term, when we modularise the panels further and start using the fledgling widget library to draw panel output, we can make use of the API to do the grunt work and introduce a set of classes for constructing Admin panels, which do the rendering in a nice OO way.

Reply:
Nice work! It would be great to separate txp `setup` functionality from extra data install. This way we could distribute "solution" packages including plugins, themes and so on.